python/pyiceberg/io/pyarrow.py

```diff
@@ -379,3 +385,62 @@ def _(_: StringType) -> pa.DataType:
 def _(_: BinaryType) -> pa.DataType:
     # Variable length by default
     return pa.binary()
+
+
+class _ConvertToArrowExpression(BoundBooleanExpressionVisitor[pc.Expression]):
+    def visit_in(self, term: BoundTerm[pc.Expression], literals: Set[Any]) -> pc.Expression:
+        return pc.field(term.ref().field.name).isin(literals)
+
+    def visit_not_in(self, term: BoundTerm[pc.Expression], literals: Set[Any]) -> pc.Expression:
+        return ~pc.field(term.ref().field.name).isin(literals)
+
+    def visit_is_nan(self, term: BoundTerm[pc.Expression]) -> pc.Expression:
+        ref = pc.field(term.ref().field.name)
+        return ref.is_null(nan_is_null=True) & ref.is_valid()
+
+    def visit_not_nan(self, term: BoundTerm[pc.Expression]) -> pc.Expression:
+        ref = pc.field(term.ref().field.name)
+        return ~(ref.is_null(nan_is_null=True) & ref.is_valid())
+
+    def visit_is_null(self, term: BoundTerm[pc.Expression]) -> pc.Expression:
```

Comment:
These bound terms also have the wrong type. I don't think that it matters for correctness, but the type of the term should be the Python type of values produced by that term, which won't be a pyarrow Expression.

Reply:
It doesn't affect the code but looks messy. Fixed in: https://github.com/apache/iceberg/pull/6308